kimchi/src/tests/foreign_field_add.rs

```diff
@@ -0,0 +1,1284 @@
+            Ok(())
+        );
+    }
+    assert_eq!(witness[8][16], -PallasField::one());
+    assert_eq!(witness[9][16], -PallasField::one());
+}
+
+#[test]
+// test when the both carries are minus one
+fn test_neg_carries() {
+    let foreign_modulus = BigUint::from_bytes_be(FOREIGN_MOD);
+    let cs = create_test_constraint_system_ffadd(1, foreign_modulus.clone());
+
+    let left_input = BigUint::from_bytes_be(OVF_NEG_BOTH);
+    let right_input = BigUint::from_bytes_be(OVF_ZERO_MI_NEG_LO);
+
+    let witness = create_witness(vec![left_input, right_input], vec![ADD], foreign_modulus);
+
+    for row in 0..=18 {
+        assert_eq!(
+            cs.gates[row].verify_witness::<Vesta>(
+                row,
+                &witness,
+                &cs,
+                &witness[0][0..cs.public].to_vec()
+            ),
+            Ok(())
+        );
+    }
+
+    for row in 16..=18 {
+        assert_eq!(
+            cs.gates[row].verify::<Vesta>(row, &witness, &cs, &[]),
+            Ok(())
+        );
+    }
+    assert_eq!(witness[8][16], -PallasField::one());
+    assert_eq!(witness[9][16], -PallasField::one());
+}
+
+#[test]
+// test the upperbound of the result
+fn test_upperbound() {
+    let foreign_modulus = BigUint::from_bytes_be(FOREIGN_MOD);
+    let cs = create_test_constraint_system_ffadd(1, foreign_modulus.clone());
+
+    let left_input = BigUint::from_bytes_be(OVF_LESS_HI_LEFT);
+    let right_input = BigUint::from_bytes_be(OVF_LESS_HI_RIGHT);
+
+    let witness = create_witness(vec![left_input, right_input], vec![ADD], foreign_modulus);
+
+    for row in 0..=18 {
+        assert_eq!(
+            cs.gates[row].verify_witness::<Vesta>(
+                row,
+                &witness,
+                &cs,
+                &witness[0][0..cs.public].to_vec()
+            ),
+            Ok(())
+        );
+    }
+
+    for row in 16..=18 {
+        assert_eq!(
+            cs.gates[row].verify::<Vesta>(row, &witness, &cs, &[]),
+            Ok(())
+        );
+    }
+}
+
+#[test]
+// test a carry that nullifies in the low limb
+fn test_null_lo_carry() {
+    let foreign_modulus = BigUint::from_bytes_be(FOREIGN_MOD);
+    let cs = create_test_constraint_system_ffadd(1, foreign_modulus.clone());
+
+    let left_input = BigUint::from_bytes_be(MAX);
+    let right_input = BigUint::from_bytes_be(NULL_CARRY_LO);
+
+    let witness = create_witness(vec![left_input, right_input], vec![ADD], foreign_modulus);
+
+    for row in 0..=18 {
+        assert_eq!(
+            cs.gates[row].verify_witness::<Vesta>(
+                row,
+                &witness,
+                &cs,
+                &witness[0][0..cs.public].to_vec()
+            ),
+            Ok(())
+        );
+    }
+
+    for row in 16..=18 {
+        assert_eq!(
+            cs.gates[row].verify::<Vesta>(row, &witness, &cs, &[]),
+            Ok(())
+        );
+    }
+    assert_eq!(witness[8][16], PallasField::zero());
+}
+
```

Comment:
You've done an amazing job with all the boundary condition tests.  This must have taken a lot of effort.

I think you should add some more test with some big random values.  Perhaps randomly generate some operands and then from these generate positive and negative test cases and verify everything is working correctly.